core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -84,19 +76,65 @@ public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpira
     deleteFiles(manifestListsToDelete, "manifest list");
   }
 
-  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
-    Set<ManifestFile> manifestFiles = Sets.newHashSet();
-    for (Snapshot snapshot : snapshots) {
-      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
-        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
-          manifestFiles.add(manifestFile.copy());
-        }
-      } catch (IOException e) {
-        throw new RuntimeIOException(
-            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
-      }
+  private Set<ManifestFile> currentManifests(
+      Set<Snapshot> snapshots, Set<ManifestFile> manifestsToDelete) {
+    Set<ManifestFile> currentManifests = ConcurrentHashMap.newKeySet();
+    if (manifestsToDelete.isEmpty()) {
+      return currentManifests;
     }
 
+    Tasks.foreach(snapshots)
+        .retry(3)
+        .stopOnFailure()
+        .throwFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (snapshot, exc) ->
+                LOG.warn(
+                    "Failed to determine manifests for snapshot {}", snapshot.snapshotId(), exc))
+        .run(
+            snapshot -> {
+              try (CloseableIterable<ManifestFile> manifestFiles = readManifests(snapshot)) {
+                for (ManifestFile manifestFile : manifestFiles) {
+                  manifestsToDelete.remove(manifestFile);
```

Comment:
Nit: Mutating `manifestsToDelete` is somewhat odd to me / is not expected from the function name & signature.

Maybe naming the method something more explicit, like `findCurrentManifestsAndRemoveFromManifestsToDelete` would be more appropriate? Or maybe breaking this into multiple functions?

If for performance reasons it's best to do it this way, then that makes sense to me.

Reply:
https://github.com/apache/iceberg/pull/5981/files#r996389820 I think my reply here should cover this, but yeah totally I think we probably need a better name here or at least inline comments